### PR TITLE
Implement draggable question for reordering

### PR DESCRIPTION
**Pull request by pkcakeout**
- Add [SortableJS](https://github.com/SortableJS/Sortable) to easily implement draggable list items. Development seems pretty active - better than "dragula", for example.
- Messily adapted the REST-API to allow patching of the order-field
- Rewriting the order of each question to force absolute ordering

I am pretty sure this is not well implemented, but maybe a nice feature! I thought: "let's try something that might be useful for grand-challenge, just to get up to speed with that project again..."

out

### CODE REVIEW

File: app/grandchallenge/reader_studies/models.py [hunk 1]
```diff
@@ -1091,11 +1091,15 @@ class Question(UUIDModel):
     class Meta:
         ordering = ("order", "created")
 
+    @property
+    def answer_type_display(self):
+        return self.get_answer_type_display()
+
```
Comment (pkcakeout) [inline app/grandchallenge/reader_studies/models.py:1097]: Added to allow use in template

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html [hunk 2]
```diff
@@ -295,11 +295,20 @@ <h2>Upload Sessions</h2>
                  aria-labelledby="v-pills-questions-tab">
                 <h2>Questions</h2>
 
-                <ul class="list-group list-group-flush">
+                <script src="https://cdn.jsdelivr.net/npm/sortablejs@latest/Sortable.min.js"></script>
+
+                <ul class="list-group list-group-flush" id="full-questions-list">
                     {% for question in object.questions.all %}
-                        <li class="list-group-item">
+                        <li class="list-group-item" gc-update-link="{% url 'api:reader-studies-question-detail' pk=question.pk %}">
```
Comment (pkcakeout) [inline app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:302]: Should I use the REST API like this? Or should I rather submit to a local API endpoint in the view?

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html [hunk 3]
```diff
@@ -295,11 +295,20 @@ <h2>Upload Sessions</h2>
                  aria-labelledby="v-pills-questions-tab">
                 <h2>Questions</h2>
 
-                <ul class="list-group list-group-flush">
+                <script src="https://cdn.jsdelivr.net/npm/sortablejs@latest/Sortable.min.js"></script>
```
Comment (pkcakeout) [inline app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:298]: Messily added this in-line. This is of course not acceptable? What is the way to do this? I saw something like this:

~~~~~.py
class Media:
    JS = ("something.js", )
    CSS = ("else.css",)
~~~~~

Is that the answer?

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html [hunk 4]
```diff
@@ -308,6 +317,36 @@ <h2>Questions</h2>
                         </li>
                     {% endfor %}
                 </ul>
+                <script>
```
Comment (pkcakeout) [inline app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:320]: Do we have a place for local JavaScript? Same pattern as jqfileuploader maybe?

File: app/grandchallenge/reader_studies/views.py [hunk 5]
```diff
@@ -896,7 +896,7 @@ def ground_truth(self, request, pk=None, case_pk=None):
         )
 
 
-class QuestionViewSet(ReadOnlyModelViewSet):
+class QuestionViewSet(ModelViewSet):
```
Comment (pkcakeout) [inline app/grandchallenge/reader_studies/views.py:899]: This seems like a brutal security issue to me. Is there a better way to allow _patching_ of the Question-model?

I only want to rewrite the "order" field of the Question-model.

File: app/grandchallenge/reader_studies/models.py (hunk 1 above)
Comment (jmsmkn) [reply app/grandchallenge/reader_studies/models.py:1097]: I'd rather keep to Django's API, you can call `{{ question.get_answer_type_display }}` from a django template, see their documentation about how template variables are resolved: https://docs.djangoproject.com/en/3.0/ref/templates/language/#variables

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html (hunk 3 above)
Comment (jmsmkn) [reply app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:298]: Yes, you're hand-rolling a form here, it's better to let django render it. Create the form in the view and add it to the context.

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html (hunk 2 above)
Comment (jmsmkn) [reply app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:302]: No, use django forms.

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html (hunk 3 above)
Comment (jmsmkn) [reply app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:298]: Also, it's best to serve it ourselves, these free CDNs are not reliable.

Comment (jmsmkn) [reply app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:298]: You can add it to packages.json.

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html [hunk 6]
```diff
@@ -308,6 +317,36 @@ <h2>Questions</h2>
...
+                    (function () {
+                        const el = document.getElementById("full-questions-list");
+
+                        function processUpdate() {
+                            let i = 1;
+                            for (const c of el.children) {
+                                const url = c.attributes.getNamedItem("gc-update-link").value;
+                                jQuery.ajax({
```
Comment (jmsmkn) [inline app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:328]: Don't use jquery ajax, use `fetch`.

File: app/grandchallenge/reader_studies/views.py (hunk 5 above)
Comment (jmsmkn) [reply app/grandchallenge/reader_studies/views.py:899]: Yes, this change bypasses all security controls that means anyone can do anything to a question.

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html (hunk 4 above)
Comment (jmsmkn) [reply app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:320]: Add it as a file in the `js` subdirectory of the reader study templates, then add it to the `script` template block.

File: app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html (hunk 2 above)
Comment (jmsmkn) [reply app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_detail.html:302]: In general, don't use custom attributes, if you need to pass extra things use data attributes: https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes

**Review by jmsmkn — CHANGES_REQUESTED** [review]
Please see my comments, I think that another approach should be taken that does not use the REST API endpoints. IMO it would be better to use a formset with the existing question forms. See https://docs.djangoproject.com/en/3.1/topics/forms/formsets/

**Comment by jmsmkn** [comment]
Looks good though! Would be nice to have this on pages too.

**Comment by pkcakeout** [comment]
Roger, I think I get it! Strategy: New sake a widget for FormSets that allows one to sort the entries instead, correct?

I will rework this to match those specs. I will convert the PR to a draft until then, because it is so far off with regards to being "mergable". Thank you for the detailed comments. It is great to learn a bit more about grand-challenge/django again!

**Comment by jmsmkn** [comment]
I'm going to close this to get it out of the list of open PRs, please feel free to re-open when you have a solution.

**Pull request closed without merging**